src/core/public/application/application_service.tsx

```diff
@@ -332,18 +334,24 @@ export class ApplicationService {
     this.currentActionMenu$.next(currentActionMenu);
   };
 
-  private async shouldNavigate(overlays: OverlayStart): Promise<boolean> {
+  private async shouldNavigate(overlays: OverlayStart, nextAppId: string): Promise<boolean> {
     const currentAppId = this.currentAppId$.value;
     if (currentAppId === undefined) {
       return true;
     }
-    const action = getLeaveAction(this.appInternalStates.get(currentAppId)?.leaveHandler);
+    const action = getLeaveAction(
+      this.appInternalStates.get(currentAppId)?.leaveHandler,
+      nextAppId
+    );
     if (isConfirmAction(action)) {
       const confirmed = await overlays.openConfirm(action.text, {
         title: action.title,
         'data-test-subj': 'appLeaveConfirmModal',
       });
       if (!confirmed) {
+        if (action.callback) {
+          action.callback();
+        }
```

Comment:
- I think this should be deferred, to avoid blocking the `navigateToApp` call if the callback is synchronous?

- This is only handling the interception when navigation to another app. If the user tries to close the tab, or to navigate to an external url, the logic is handled in `onBeforeUnload`. In that case, we definitely want to defer the call to `action.callback` as any 'too long' synchronous handler will be ignored by the browser.